File: 3-automation-Playwright-Hoyts-Shop/pages/pageObjects/mainPageHoytsShop.ts
import { Page, Locator } from "@playwright/test";
import { abstractComponentsHoytsShop } from "../abstractComponents/abstractComponentsHoytsShop";
import { error } from "node:console";
export class mainPageHoytsShop extends abstractComponentsHoytsShop {

    constructor(page: Page){
        super(page);
    }
    // ========================= Action Methods =======================

    // == Go to URL ===================================================
    async goTo(){
        await this.page.goto('https://shop.hoyts.com.au/', {waitUntil: "load", timeout: 15000});
    }
    // == Carousel =====================================================
    async scrollInCarousel(title: string, carousel: string){
        const viewport = await this.page.evaluate(() => window.innerWidth);
        if(viewport > 767) return;
        let sectionTitle ='';

        if(carousel === 'gift card'){
            sectionTitle = "Most Popular Gift Cards";
        }else if(carousel === 'voucher'){
            sectionTitle = "Most Popular Vouchers";
        }

        const section = this.page.locator(`//section[.//header[contains(., "${sectionTitle}")]]`);
        await section.waitFor({state : "visible"});

        const product = section
            .locator('.keen-slider a[title]')
            .filter({ hasText: title });

        const rightArrow = section.locator('.keen-slider__arrow--right');

        for (let i = 0; i < 10; i++) {
            const giftCard = product.first();

            if (await giftCard.count() > 0) {
            const isInViewport = await giftCard.evaluate((el) => {
                const rect = el.getBoundingClientRect();
                return rect.left >= 0 && rect.right <= window.innerWidth;
            });

            if (isInViewport) {
                await giftCard.scrollIntoViewIfNeeded();
                return;
            }
            }

            await rightArrow.click();
            await this.page.waitForTimeout(300);
        }

        throw new Error(`Gift Card: ${title} not available`);

    }

    // ========================= Pick Most Popular Gift Cards =======================
    async pickMostPopularGiftCard( title: string, value: string){
        await this.page.evaluate(()=>window.scrollBy(0,100));
        await this.scrollInCarousel(title, 'gift card');
        const container = this.productContainer(title);
        if(await container.count() ===0){
            throw new Error (`Gift Card '${title}' not found in the most popular gift cards`);
        }

        await container.first().waitFor({ state: 'visible', timeout: 5000 });
        await this.selectGiftCard(container, value);
    }

}

File: 3-automation-Selenium-Hoyts-Shop/src/main/java/hoytsShopPageObjects/mainPageHoytsShop.java
package hoytsShopPageObjects;

import org.openqa.selenium.*;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;

import abstractComponents.hoytsShopAbstractComponents;

public class mainPageHoytsShop extends hoytsShopAbstractComponents {

    public mainPageHoytsShop(WebDriver driver) {
        super(driver);
        PageFactory.initElements(driver, this);
    }
    
  
    
  // ========================= Action Methods =======================

 // == Go to Website ================================================
    
    public void goTo() {
        driver.get("https://shop.hoyts.com.au/");
    }
    
// == Scroll to the Gift Card in Carousel (if applicable) =========================
    
    public void scrollInCarousel(String title, String carousel) {
    	int viewport = ((Long)((JavascriptExecutor)driver).executeScript("return window.innerWidth")).intValue();
    	if (viewport > 767) return;
     
    	String sectionTitle = "";
    	if (carousel.equalsIgnoreCase("gift card")) {
    		sectionTitle = "Most Popular Gift Cards";  // For Gift Cards
    	}
    	else if (carousel.equalsIgnoreCase("voucher")) {
            sectionTitle = "Most Popular Vouchers";  // For Vouchers
    	}
    	
    	WebElement section = waitForElementToBeVisible(driver.findElement(By.xpath(
    			"//section[.//header[contains(.,'" + sectionTitle + "')]]")));
    	((JavascriptExecutor)driver).executeScript("arguments[0].scrollIntoView({block: 'center'});", section);
    	
    	WebElement slider = section.findElement(By.cssSelector(".keen-slider"));
    	WebElement rightArrow = section.findElement(By.cssSelector(".keen-slider__arrow--right"));
    	WebElement leftArrow = section.findElement(By.cssSelector(".keen-slider__arrow--left"));
    	
    	By productLocator = By.xpath(".//a[contains(@title, '"+ title +"')]");
    	
    	for (int i = 0; i < 10; i++) {

    	    if (slider.findElements(productLocator).size() > 0) {
    	        WebElement product = slider.findElement(productLocator);

    	        if (product.isDisplayed()) {

    	            ((JavascriptExecutor) driver)
    	                .executeScript("arguments[0].scrollIntoView({block:'center', inline:'center'});", product);

    	            return;
    	        }
    	    }

    	    ((JavascriptExecutor) driver)
    	        .executeScript("arguments[0].click();", rightArrow);

    	    try { Thread.sleep(500); } catch (InterruptedException e) {}
    	}
    	throw new RuntimeException("Product: "+ title + " not available");
    }


// == Pick Most Popular Gift Cards ===================================
    public void pickPopularGiftCard(String title, String value) {

        ((JavascriptExecutor) driver)
                .executeScript("window.scrollBy(0,100)");
        
        scrollInCarousel(title, "gift card");

        By containerLocator = productContainer(title);

        if (driver.findElements(containerLocator).size() == 0) {
            throw new RuntimeException("Gift Card '" + title + "' not found in most popular");
        }

        WebElement container = waitForElementToBeVisible(driver.findElement(containerLocator));

        selectGiftCard(container, value);
    }
}